### PR TITLE
Fixed #15760 -- Added JavaScript hooks for admin inlines

### PR DESCRIPTION
**Pull request by RamezIssac**
Ticket: https://code.djangoproject.com/ticket/15760

### CODE REVIEW

**Comment by charettes** [comment]
Is there a reason were not firing a custom event event instead? e.g. `'formset:added'` and `'formset:removed'`.

What if two third party application need to hook to those callbacks?

**Comment by RamezIssac** [comment]
Nope no reason actually.

This sound a better idea especially considering the 2 apps scenario.

Will look into that path.

**Comment by mlavin** [comment]
My original proposal/patch was to use events as well as #542 which was created for this ticket.

**Comment by timgraham** [comment]
Now that JavaScript linting and unit testing has been added in 2d0dead224b6448072b72b37d2fbcc8dc3afa007, could you rebase and update the patch? After you do so, please uncheck "Patch needs improvement" so the ticket appears in the review queue.

File: docs/ref/contrib/admin/javascript.txt [hunk 1]
```diff
@@ -0,0 +1,74 @@
+
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: no blank line

File: docs/ref/contrib/admin/javascript.txt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+========================
+Admin javascript tweaks
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: JavaScript

File: docs/ref/contrib/admin/javascript.txt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+========================
+
+.. versionadded:: developement
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: development -> 1.9

File: docs/ref/contrib/admin/javascript.txt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+========================
+Admin javascript tweaks
+========================
+
+.. versionadded:: developement 
+
+Inline hooks
+------------
+
+Sometimes you need to execute some JavaScript or jQuery upon adding or removing inlines in the admin change_form.
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: chop "or jQuery" (that's a subset of JavaScript)
Please wrap lines at 79 characters and use backticks around code-type things like variable names, etc.

File: docs/ref/contrib/admin/javascript.txt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+Django admin provide two jQuery events formset:added and formset:removed to do so.
+The event handler is passed three arguments event, $row, formsetName.
+event: is the ``jQuery`` event.
+$row: is the newly added (or removed) row.
+formsetName: the the formset the row belongs to.
+
+The event is fired using the ``django.jQuery`` namespace. See :doc:`/ref/contrib/admin/jQuery` for information.
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: You need to add `.. _contrib-admin-jquery` above the jQuery heading on the other page, then use
```
:ref:`contrib-admin-jquery`
```
here.

File: docs/ref/contrib/admin/javascript.txt [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+In your custom ``change_form.html``, extends ``admin_change_form_document_ready`` and add 
+the events' listener code like this ::
+
+    {% extends 'admin/change_form.html' %}
+    {% block admin_change_form_document_ready %}
+    {{ block.super }}
+    <script>
+            (function ($) {
+                $(document).on('formset:added', function (event,$row, formsetName) {
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: no space after function
space after commas

File: docs/ref/contrib/admin/javascript.txt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                    if(formsetName == 'author_set'){
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: space after if and before {

File: docs/ref/contrib/admin/javascript.txt [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+The event is fired using the ``django.jQuery`` namespace. See :doc:`/ref/contrib/admin/jQuery` for information.
+
+In your custom ``change_form.html``, extends ``admin_change_form_document_ready`` and add 
+the events' listener code like this ::
+
+    {% extends 'admin/change_form.html' %}
+    {% block admin_change_form_document_ready %}
+    {{ block.super }}
+    <script>
+            (function ($) {
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: indent 4 spaces instead of 8?

**Comment by RamezIssac** [comment]
Dear Tim , 
Is this pull awaiting review ? or there is still any problems with it ?
Did i edited authors too soon or something ?! :-)
Any review is highly appreciated.

Kind Regards;

File: docs/ref/contrib/admin/javascript.txt (hunk 1 above)
Comment (RamezIssac) [inline docs/ref/contrib/admin/javascript.txt]: The blank line was at the very beginning of file when you made the line note, i removed it,  but the line note continues scrolling down to each new blank line in the file.

File: django/contrib/admin/static/admin/js/inlines.js [hunk 9]
```diff
@@ -97,6 +97,7 @@
                     // If a post-delete callback was provided, call it with the deleted form:
                     if (options.removed) {
                         options.removed(row);
+                        $(document).trigger('formset:removed', [row, options.prefix]);
```
Comment (mlavin) [inline django/contrib/admin/static/admin/js/inlines.js]: These events should be triggered regardless of whether the callback is given. The use case here is that a third party plugin can't specify the callback and that is why they need the event to fire.

Comment (RamezIssac) [inline django/contrib/admin/static/admin/js/inlines.js]: Ok, but just for clarification, these callback `options.removed`, `options.added` are always given, as they are a part of Django admin Tabular & Stacked inline. (inlines.js line:199 & 265).

**Comment by timgraham** [comment]
Could you see if it's possible to use the new JavaScript unit test framework `js_tests/admin/inline.tests.js` to test this? Selenium tests have always been rather unpleasant to maintain and I think a unit test could be much simpler. Let me know what you think.

File: docs/ref/contrib/admin/javascript.txt [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+Django admin provide two jQuery events ``formset:added`` and ``formset:removed``.
+The event handler is passed three arguments ``event``, ``$row``, ``formsetName``.
+
+``event``: is the ``jQuery`` event.
+
+``$row``: is the newly added (or removed) row.
+
+``formsetName``: the the formset the row belongs to.
+
+The event is fired using the ``django.jQuery`` namespace. See :ref:`contrib-admin-jquery` for information.
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: please wrap docs lines at 79 chars

File: docs/ref/contrib/admin/javascript.txt [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    </script>
+    {% endblock %}
+    
+
+* The JavaScript code have to go in a block if you are inhering Django change_form.html or it wont be rendered in the final html.
+* We added ``{{ block.super }}`` because Django's `admin_change_form_document_ready` block 
+  contains JavaScript code to handle various operations in the change_form, and we need that to be rendered too.
+
+Sometimes you'll need to work with ``jQuery`` plugins that are not registered in the django.jQuery namespace.
+To do that, simply change how the code listen for event and instead of warping the whole logic in django.jQuery namespace,
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: warping ->wrapping?

Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: listen for event -> listens for events

File: docs/ref/contrib/admin/javascript.txt [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+The event is fired using the ``django.jQuery`` namespace. See :ref:`contrib-admin-jquery` for information.
+
+In your custom ``change_form.html``, extends block ``admin_change_form_document_ready`` and add 
+the events' listener code like this ::
+
+    {% extends 'admin/change_form.html' %}
+    
+    {% block admin_change_form_document_ready %}
+    {{ block.super }}
+    <script>
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: type="text/javascript"?

File: docs/ref/contrib/admin/javascript.txt [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    (function($){
+        $(document).on('formset:added', function(event, $row, formsetName) {
+            if (formsetName == 'author_set') {
+                //Do something
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: consistent spacing after //

File: docs/ref/contrib/admin/javascript.txt [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+Sometimes you'll need to work with ``jQuery`` plugins that are not registered in the django.jQuery namespace.
+To do that, simply change how the code listen for event and instead of warping the whole logic in django.jQuery namespace, 
+just listen to the event from there. ::
+
+    {% extends 'admin/change_form.html' %}
+    
+    {% block admin_change_form_document_ready %}
+    {{ block.super }}
+    <script>
+    function($){
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: is `function($)` correct - that function is never invoked as far as I can see?

File: docs/ref/contrib/admin/javascript.txt [hunk 15]
```diff
@@ -0,0 +1,76 @@
+===========
+JavaScript
+===========
+.. versionadded:: 1.9 
+
+Inline hooks
+------------
+
+Sometimes you need to execute some JavaScript upon adding or removing inlines 
+in the admin change_form.
```
Comment (timgraham) [inline docs/ref/contrib/admin/javascript.txt]: change form

**Comment by timgraham** [comment]
Sorry for the delayed feedback. I'll try to give the docs a bit more polish after you next round of updates. Thanks!

**Comment by timgraham** [comment]
Did you see my comment about trying for a JavaScript test instead of selenium or did you just not have time to look at it yet?

**Comment by RamezIssac** [comment]
Yes i did, and yes still didn't get the time to look at the JavaScript
testing yet... Hopefully will soon.

**Comment by RamezIssac** [comment]
Thank you for the great support Tim.
I added JavaScript Tests and added an indication of its existence in the 'normal' test doc string

File: js_tests/admin/inlines.test.js [hunk 16]
```diff
@@ -26,3 +26,18 @@ test('add form', function(assert) {
     addButton.click();
     assert.ok(this.table.find('#first-1').hasClass('row2'));
 });
+
+test('inline hooks', function(assert) {
+    assert.expect(2);
+    var $document = django.jQuery(document);
+    var addButton = this.table.find('.add-row a');
+    $document.on( "formset:added", function(event, $row, formsetName) {
```
Comment (timgraham) [inline js_tests/admin/inlines.test.js]: Why a space after ( on this line and the next?
Can we verify the arguments in this test too? Would be nice if we could get away without adding the selenium test in my opinion

**Comment by timgraham** [comment]
doc edits: http://dpaste.com/3T9BEDZ
A mention in the 1.9 release notes is appropriate too.

**Comment by RamezIssac** [comment]
Got it. Thanks.
Will update the pull in few hours.
On Sep 20, 2015 12:56 AM, "Tim Graham" <notifications@github.com> wrote:

doc edits: http://dpaste.com/3T9BEDZ
A mention in the 1.9 release notes is appropriate too.

—
Reply to this email directly or view it on GitHub
<https://github.com/django/django/pull/4883#issuecomment-141712607>.

**Comment by RamezIssac** [comment]
Applied docs diff, rolled-back selenium test and rely entirely on JavaScript test, added 1.9 release note.
Something wired with the 'default' test results though,  its complains are not related to this pull. Would you kindly check it and 're-test' maybe ?! Thanks.

**Comment by timgraham** [comment]
Merged in 1335aa2fb9f590d786a5baee1b781377dc3a2ec8, thanks!

**Pull request closed without merging**